### DISCUSSION TITLE
Vue-like Single File Components

### DISCUSSION THREAD

**Discussion by rinodrummer** [livewire/livewire#1944] [3. Feature ideas]
Hi guys, I'm extremely new with the framework and I really love it, but I find a bit uncomfortable with the fact that I should use 2 files for one component (class and view) or I have to use Inline Components that uses heredoc that I find extremely untidy.

I come from a Vue experience that offers a good separation of concerns, using the same file.

It would really awesome to have a `Counter.wire` file with kinda these structures:

<details>
<summary>Class-like structure</summary>

```php
<?php
    class Counter extends Component {
        // ...
        
        // I don't know if `render()` method is needed anymore.
    }

    @section('view')
        // This is the main section, it gets rendered by default.
    @endsection
```
</details>

<details>
<summary>Blade-like structure</summary>

```blade
@php
    class Counter extends Component {
        // ...
        
        // I don't know if `render()` method is needed anymore.
    }
@endphp

@section('view')
    {{-- This is the main section, if not present, compilation should fail. --}}
@endsection
```
</details>

It will lead to a lot of features, like the definition of sub-section that will help to ensure an even bigger SoC:

<details>
<summary>Multiple sections in the same file</summary>

```blade
@php
    class Counter extends Component {
        public $count = 0;

        public function increment()
        {
            $this->count++;
        }
        
        public function decrement()
        {
            $this->count++;
        }
    }
@endphp

@section('view')
    <div>
        <p>Current value: {{ $count }}</p>
        
        @yield('increment-button')
        @yield('decrement-button')
    </div>
@endsection

@section('increment-button')
    <button wire:click="increment">+</button>
@endsection

@section('decrement-button')
    <button wire:click="decrement">-</button>
@endsection
```
</details>

<details>
<summary>Auto-wrapping view code when multiple root elements</summary>

Having multiple root elements causes problems to Livewire components.
Doing in this way, we don't solve the problem, but handle it an very simple way.

<details>
<summary>String-typed value for <code>$wrap</code></summary>

```blade
@php
    class Counter extends Component {
        public $count = 0;
        
        /**
         * Specifies the tag which will wrap the view section.
         *
         * @var string Tag name. Default value is 'div'.
         */
        protected $wrap = 'section';

        public function increment()
        {
            $this->count++;
        }
        
        public function decrement()
        {
            $this->count++;
        }
    }
@endphp

@section('view')
    <p>Current value: {{ $count }}</p>
        
    @yield('increment-button')
    @yield('decrement-button')
@endsection

@section('increment-button')
    <button wire:click="increment">+</button>
@endsection

@section('decrement-button')
    <button wire:click="decrement">-</button>
@endsection
```

Output:

```html
<section>
    <p>Current value: 0</p>
    
    <button wire:click="increment">+</button>
    <button wire:click="decrement">-</button>
</section>
```
</details>

<details>
<summary>Array-typed value for <code>$wrap</code></summary>

```blade
@php
    class Counter extends Component {
        public $count = 0;
        
        protected $wrap = [
            'tag' => 'section',
            'attributes' => [
                'class' => 'text-center',
                'id' => 'counter'
            ]
        ];

        public function increment()
        {
            $this->count++;
        }
        
        public function decrement()
        {
            $this->count++;
        }
    }
@endphp

@section('view')
    <p>Current value: {{ $count }}</p>
        
    @yield('increment-button')
    @yield('decrement-button')
@endsection

@section('increment-button')
    <button wire:click="increment">+</button>
@endsection

@section('decrement-button')
    <button wire:click="decrement">-</button>
@endsection
```

Output:

```html
<section class="text-center" id="counter">
    <p>Current value: 0</p>
    
    <button wire:click="increment">+</button>
    <button wire:click="decrement">-</button>
</section>
```
</details>
</details>

I know that Vue uses a DOM structure that respect the nature of an HTML document, so it doesn't break things meanwhile this kind of request needs a not-so-small kind of effort because it must introduce a compiled routine (kinda like Blade), but I think that it may have really good impact on the realization of Livewire Components.

I know that maybe you all already read this idea (I tried to search it on GitHub, but I haven't found it), that my request is far away from the best structured and my english may lack in some points, but please, take this suggestion like a demonstration of big admiration to the project and every contributor who works on it.

Let me know your opinion on it!
Thanks in advice!

**Comment by nuernbergerA** [comment]
What would be the difference between `Counter.wire` and `Couter.php` (with blade heredoc inside render)?

**Reply by rinodrummer** [reply]
One the biggest problems I have with heredoc is that using it in plain PHP tends to mix HTML/Blade code with it, making code harder to read.

The only advantage (which I personally find very useful) is that we have a better separation of concerns.

I know that my request adds a complexity in the whole system, but will surely ensure a better approach to the code.
